### PR TITLE
Use a dedicated chip-types.xml file into src/app/zap-templates/zcl

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem
 The zcl types from `src/app/zap-templates/zcl/data-model/silabs/types/` does not fully match the types from the spec, and there is also a lot of clusters into `src/app/zap-templates/zcl/data-model/silabs` that are not needed at that point. In order to try to align a bit more with the spec, this PR shuffle a bit `src/app/zap-templates/zcl` in order to makes it easier to spot the differences.

#### Change overview
 * Update types from the xml definition. Update the code accordingly where it does not compile anymore
 * Add `src/app/zap-templates/zcl/data-model/chip` to host definition written for Chip
 * Remove a whole set of definitions from `src/app/zap-templates/zcl/data-model/Silabs`

#### Testing
 It was tested by running locally `scripts/tests/test_suites.sh` once everything has been updated.

### CODE REVIEW

**Comment by vivien-apple** [comment]
One thing that I have note that from the spec that I'm not sure of is the fact that `ClusterId` is now a `uint32` , as well as `AttributeId`.

**Review by msandstedt — APPROVED** [review]

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp [hunk 1]
```diff
@@ -6380,8 +6380,8 @@ void DispatchServerCommand(app::Command * apCommandObj, CommandId aCommandId, En
 void DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aCommandId, chip::EndpointId aEndPointId,
                                   chip::TLV::TLVReader & aReader, Command * apCommandObj)
 {
-    ChipLogDetail(Zcl, "Received Cluster Command: Cluster=%" PRIx16 " Command=%" PRIx8 " Endpoint=%" PRIx8, aClusterId, aCommandId,
-                  aEndPointId);
+    ChipLogDetail(Zcl, "Received Cluster Command: Cluster=%" PRIx32 " Command=%" PRIx32 " Endpoint=%" PRIx16, aClusterId,
```
Comment (tima-q) [inline examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:6383]: cluster ID's are 16-bit - commands and endpoints 8-bit 
- see https://github.com/project-chip/connectedhomeip/blob/ed677f04ae8187098130f3701b5583f8db56a55e/src/app/util/basic-types.h#L33

Why the change into 32-bit/16-bit prints ?

Comment (tima-q) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:6383]: > One thing that I have note that from the spec that I'm not sure of is the fact that `ClusterId` is now a `uint32` , as well as `AttributeId`.
Sorry - now read your comment here. 
I expect 32-bit would create an significant code/frame size impact. 
@tcarmelveilleux Would you know the spec reasoning for these sizes ?

Comment (vivien-apple) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:6383]: > One thing that I have note that from the spec that I'm not sure of is the fact that `ClusterId` is now a `uint32` , as well as `AttributeId`.

I asked around to double check this is not a typo in the spec. The move to `uint32` for `Attribute Id`, `Command Id` and `Cluster Id` is expected. The move of `uint16` for `Endpoint Id` is expected too.

**Review by woody-apple — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp (hunk 1 above)
Comment (tima-q) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:6383]: I do see the 32-bit is actually a combination of: 16-bit 'Namespace ID' | 16-bit Cluster ID. (10.6.2.5 - v0.7 ballot version)
It's noted that for standard CHIP clusters the Cluster ID can compress down to 16-bit. (namespace being all 0x0000 in that case)
Is it an option to handle these vendor specific clusters separately ? The bulk of the cluster handling isn't using the top part probably.

File: src/app/util/basic-types.h [hunk 2]
```diff
@@ -28,17 +28,24 @@
...
 
+// Pull in VendorId
+#include <core/CHIPVendorIdentifiers.hpp>
+
 namespace chip {
-typedef uint8_t EndpointId;
-typedef uint16_t ClusterId;
-typedef uint16_t AttributeId;
+typedef uint32_t ClusterId;
+typedef uint32_t AttributeId;
```
[the remark this answers is not in the document]
Comment (vivien-apple) [reply src/app/util/basic-types.h:36]: > This is a little dubious because we're still using 16-bit attribute ids in the ReadAttributes and ReadAttributesResponse commands , no? I am not seeing that changed in `process-global-message.cpp`.
> 
> I'm actually a bit surprised this compiled, since we are passing `attrId` values to `emberAfPutInt16uInResp`. I guess we are not using `-Wconversion` in any of the places that build the process-global-message code? I'll look into fixing that....

IIRC we used to build with `-Wconversion` into various places, such as chip-tool, the build system shift to use the gn template `chip_data_model.gni` may have got rid of `src/app` beeing under the `-Wconversion` radar.

> 
> Given our lack of test/build coverage there, I'm really worried about making these size changes. It might be OK, but it might also make turning on `-Wconversion` pretty hellish....
> 
> At the very least, I would suggest leaving `AttributeId` specifically as 16-bit for now and then making that change in a separate PR after we turn on `-Wconversion` for most of this code.

I locally rebased my PR on top of #7902 and fixed all the places where `-WConversion` was complaining as well as some places where it was not complaining because of hand written cast. 
I did run `./scripts/tests/test_suites.sh`, did some tests for for discover attributes and attributes reporting after all those modifications, so I think that will be OK to keep `AttributeId` with those modifications.

**Comment by vivien-apple** [comment]
In order to make it easier to properly review this PR, I'm going to split it into 2 parts. The first one will contains the removals of clusters (#7915) while this one will focus on updating the types and the various callsites.

**Comment by vivien-apple** [comment]
I pushed an update with only the types changes. If needed and in order to make it easier to review or to avoid regressions I can split it into multiple patches. One per type.

**Comment by woody-apple** [comment]
@mspang @Damian-Nordic @jmartinez-silabs @LuDuda

@vivien-apple conflicts here

File: src/app/zap-templates/zcl/data-model/chip/chip-types.xml [hunk 3]
```diff
@@ -53,39 +41,37 @@ limitations under the License.
...
     <type id="0xF1" description="Vendor ID"                name="vendor_id"         size="2"  discrete="true" />
-    <type id="0xF2" description="Device Type ID"           name="device_type_id"    size="4"  discrete="true" />
+    <type id="0xF2" description="Device Type ID"           name="devtype_id"        size="4"  discrete="true" />
     <type id="0xF3" description="Fabric ID"                name="fabric_id"         size="8"  discrete="true" />
     <type id="0xF4" description="Group ID"                 name="group_id"          size="2"  discrete="true" />
     <type id="0xF5" description="Status Code"              name="status"            size="4"  discrete="true" />
     <type id="0xF6" description="Data Version"             name="data_ver"          size="4"  discrete="true" />
     <type id="0xF7" description="Event Number"             name="event_no"          size="8"  discrete="true" />
-    <type id="0xF8" description="Endpoint Number"          name="endpoint_id"       size="1"  discrete="true" />
+    <type id="0xF8" description="Endpoint Number"          name="endpoint_no"       size="2"  discrete="true" />
```
Comment (jmartinez-silabs) [inline src/app/zap-templates/zcl/data-model/chip/chip-types.xml:74]: Don't we usually call it Endpoint ID ?

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/util/attribute-storage.cpp [hunk 4]
```diff
@@ -893,19 +893,19 @@ static uint8_t findClusterEndpointIndex(EndpointId endpoint, ClusterId clusterId
...
+                                                                                              mask, manufacturerCode) != NULL)
+                                         ? 1
+                                         : 0));
     }
 
     return epi;
 }
 
-static uint8_t findIndexFromEndpoint(EndpointId endpoint, bool ignoreDisabledEndpoints)
+static uint16_t findIndexFromEndpoint(EndpointId endpoint, bool ignoreDisabledEndpoints)
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-storage.cpp:906]: At some point, maybe we should have an EndpointIndex type?   Followup, presumably.

File: src/app/util/attribute-storage.cpp [hunk 5]
```diff
@@ -919,12 +919,12 @@ static uint8_t findIndexFromEndpoint(EndpointId endpoint, bool ignoreDisabledEnd
 
 bool emberAfEndpointIsEnabled(EndpointId endpoint)
 {
-    uint8_t index = findIndexFromEndpoint(endpoint,
-                                          false); // ignore disabled endpoints?
+    uint16_t index = findIndexFromEndpoint(endpoint,
+                                           false); // ignore disabled endpoints?
 
-    EMBER_TEST_ASSERT(0xFF != index);
+    EMBER_TEST_ASSERT(0xFFFF != index);
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-storage.cpp:925]: And a named constant for "invalid index" or something....  Again, followup.

File: src/app/util/process-global-message.cpp [hunk 6]
```diff
@@ -642,12 +641,12 @@ bool emAfProcessGlobalCommand(EmberAfClusterCommand * cmd)
...
         if (zclCmd == ZCL_DISCOVER_COMMANDS_RECEIVED_COMMAND_ID)
         {
-            emberAfPutInt8uInResp(ZCL_DISCOVER_COMMANDS_RECEIVED_RESPONSE_COMMAND_ID);
+            emberAfPutInt32uInResp(ZCL_DISCOVER_COMMANDS_RECEIVED_RESPONSE_COMMAND_ID);
             flag = false;
         }
         else
         {
-            emberAfPutInt8uInResp(ZCL_DISCOVER_COMMANDS_GENERATED_RESPONSE_COMMAND_ID);
+            emberAfPutInt32uInResp(ZCL_DISCOVER_COMMANDS_GENERATED_RESPONSE_COMMAND_ID);
```
Comment (bzbarsky-apple) [inline src/app/util/process-global-message.cpp:649]: I really wish we had added an emberAfPutCommandIdInResp and same for reading and for other types...  but since we plan to switch to TLV from this stuff anyway, probably not worth it.

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp (hunk 1 above)
Comment (tima-q) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:6503]: A question for follow-up.
The current ZCL handling code already seems to have a fully separate code path for manufacturer specific cluster handling.
It would avoid using the 32-bit in 90% of the cases where the top 16-bit is all 0x0000 = a standard cluster.

Comment (bzbarsky-apple) [reply examples/all-clusters-app/all-clusters-common/gen/IMClusterCommandHandler.cpp:6503]: That's because in ZCL you can't have manufacturer-specific granularity finer than "entire cluster"....

File: src/app/zap-templates/zcl/data-model/chip/chip-types.xml (hunk 3 above)
Comment (vivien-apple) [reply src/app/zap-templates/zcl/data-model/chip/chip-types.xml:74]: In the spec, it is named `endpoint-no`. I tried to follow the spec naming.

File: examples/all-clusters-app/all-clusters-common/gen/attribute-size.cpp [hunk 7]
```diff
@@ -195,14 +195,14 @@ uint16_t emberAfCopyList(ClusterId clusterId, EmberAfAttributeMetadata * am, boo
...
-                           sizeof(entry->type)); // DEVICE_TYPE_ID
+                           sizeof(entry->type)); // DEVTYPE_ID
             copyListMember(write ? dest : (uint8_t *) &entry->revision, write ? (uint8_t *) &entry->revision : src, write,
                            &entryOffset, sizeof(entry->revision)); // INT16U
             break;
         }
         case 0x0001: // server list
         {
-            entryLength = 2;
+            entryLength = 4;
```
[the remark this answers is not in the document]
Comment (vivien-apple) [reply examples/all-clusters-app/all-clusters-common/gen/attribute-size.cpp:205]: This is autogenerated code. I agree it would be nice if we could use `sizeof` but since attributes can also stored `struct` members that won't work :(

[the remark this answers is not in the document]
Comment (tima-q) [reply examples/all-clusters-app/all-clusters-common/gen/attribute-size.cpp:205]: too bad - or are the structs perhaps defined with `__attribute__((packed))` ? :)

File: examples/all-clusters-app/all-clusters-common/gen/attribute-size.cpp [hunk 8]
```diff
@@ -952,7 +952,7 @@ uint16_t emberAfAttributeValueListSize(ClusterId clusterId, AttributeId attribut
     uint32_t totalSize = kSizeLengthInBytes + (entryCount * entryLength);
     if (!chip::CanCastTo<uint16_t>(totalSize))
     {
-        ChipLogError(Zcl, "Cluster 0x%04x: Size of attribute 0x%02x is too large.", clusterId, attributeId);
+        ChipLogError(Zcl, "Cluster %" PRIx32 ": Size of attribute %" PRIx32 " is too large.", clusterId, attributeId);
```
[the remark this answers is not in the document]
Comment (vivien-apple) [reply examples/all-clusters-app/all-clusters-common/gen/attribute-size.cpp:955]: All the different versions of printing commands, attributes, ... have also annoy me a lot while doing this patch. I was thinking of adding/generating macros such as `ChipLogFormatX64` in a followup.

So for example there could be a `ChipLogFormatCommandId`,  `ChipLogFormatAttributeId`, etc.. That way all formatting will be the same across the tree, and we could easily print the manufacturer code and the command code separately.
